### PR TITLE
Ensure the namespaces filtering is respected in packages.config package installation/updates in PMUI

### LINKED ISSUE #10738: Ensure the namespaces filtering is respected in packages.config package installation/updates in PMUI
In PR, restore and package installation are equivalent. 
In PC they are not. 

This task covers ensuring that the namespaces are respected during package installation. 

*Detailed explanation:*

* In https://github.com/NuGet/Home/issues/10823, support for namespaces filtering was added in restore. 
* This issue ensures that commands run through:
    * Package Manager UI, both project level and solution level (I don't expect code differences here, just being explicit)
   
~~Please consider splitting up the work in smaller tasks whenever possible to allow for easier reviewing~~
I created separate issues for:
   * PMC, any commands. https://github.com/NuGet/Home/issues/11001
    * Any IVs installer APIs. https://github.com/NuGet/Home/issues/11002

### PR DESCRIPTION
**Pull request by erdembayar**
## Bug

Fixes: https://github.com/NuGet/Home/issues/10738

Regression? Last working version: N/A

## Description
This change ensuring that the namespaces are respected during package installation PMUI(solution and project). If there is same package Id exist in 2 repositories then it only request it from one it specified in namespace filter, previously we try to fetch from all possible repositories and install one from fastest which may not be what user really wanted.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [ ] N/A

### CODE REVIEW

File: test/EndToEnd/tests/PackageNameSpaceTests.ps1 [hunk 1]
```diff
@@ -115,6 +115,117 @@ function Test-PackageNamespaceRestore-WithMultipleFeedsWithIdenticalPackages-Res
...
+        $projectDirectoryPath = $p.Properties.Item("FullPath").Value
+        $solutionDirectory = Split-Path -Path $projectDirectoryPath -Parent   
+
+        CreateCustomTestPackage "Contoso.MVC.ASP" "1.0.0" $privateRepo "Thisisfromprivaterepo.txt"
+        CreateCustomTestPackage "Contoso.MVC.ASP" "2.0.0" $privateRepo "Thisisfromprivaterepo2.txt"
+        CreateCustomTestPackage "Contoso.MVC.ASP" "1.0.0" $opensourceRepo "Thisisfromopensourcerepo.txt"
+        CreateCustomTestPackage "Contoso.MVC.ASP" "2.0.0" $opensourceRepo "Thisisfromopensourcerepo2.txt"
+
+        # Act
+        Install-Package "Contoso.MVC.ASP" -ProjectName $p.Name -version 1.0.0
```
Comment (erdembayar) [inline test/EndToEnd/tests/PackageNameSpaceTests.ps1]: Not sure if it's installing from Solution level PMUI or project level PMUI. 
Please let me know if this new tests are not enough.

Comment (nkolev92) [reply test/EndToEnd/tests/PackageNameSpaceTests.ps1]: The install-package command that's used in End to end tests is the PMC one. 

To add a PM UI specific test, you'd need to use apex tests instead which know how to call the PM UI instead. https://github.com/NuGet/NuGet.Client/blob/1d75910076b2ecfbe5f142227cfb4fb45c093a1e/test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs#L49

Good news is that your change seems to affect *all* of packages.config in general, so it shouldn't be too bad :)

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs [hunk 2]
```diff
@@ -1736,8 +1736,9 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
                     var primaryPackages = new List<PackageIdentity> { packageIdentity };
 
                     HashSet<SourcePackageDependencyInfo> availablePackageDependencyInfoWithSourceSet = null;
+                    PackageNamespacesConfiguration packageNamespacesConfiguration = PackageNamespacesConfiguration.GetPackageNamespacesConfiguration(Settings);
 
-                    var gatherContext = new GatherContext()
+                    var gatherContext = new GatherContext(packageNamespacesConfiguration)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:1743]: I think there's a good chance to add some tests for the NuGetPackageManager class as well.

File: src/NuGet.Core/NuGet.PackageManagement/Resolution/GatherContext.cs [hunk 3]
```diff
@@ -74,6 +80,8 @@ public GatherContext()
         /// </summary>
         public bool IsUpdateAll { get; set; }
 
+        public PackageNamespacesConfiguration PackageNamespacesConfiguration { get; }
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.PackageManagement/Resolution/GatherContext.cs:86]: Add docs please. 
Call out that it may be null.

File: src/NuGet.Core/NuGet.PackageManagement/Resolution/ResolverGather.cs [hunk 4]
```diff
@@ -533,13 +535,37 @@ private void QueueWork(IReadOnlyList<SourceResource> sources, string packageId,
 
         private void QueueWork(IReadOnlyList<SourceResource> sources, PackageIdentity package, bool ignoreExceptions, bool isInstalledPackage)
         {
+            IReadOnlyList<string> configuredPackageSources = null;
+
+            if (_isPackageNamespaceEnabled)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.PackageManagement/Resolution/ResolverGather.cs]: You can add tests for this in https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.PackageManagement.Test/ResolverGatherTests.cs as well.

**Review by nkolev92 — COMMENTED** [review]
Glad you made progress narrowing down a change that would likely work *everywhere*. 

The trick with that is we'd need to ensure that *every* component actually does work the way we'd expect it to.

File: src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs (hunk 2 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:1743]: Added.

File: src/NuGet.Core/NuGet.PackageManagement/Resolution/GatherContext.cs (hunk 3 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.PackageManagement/Resolution/GatherContext.cs:86]: Addressed.

File: src/NuGet.Core/NuGet.PackageManagement/Resolution/ResolverGather.cs (hunk 4 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.PackageManagement/Resolution/ResolverGather.cs]: Added, please check.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs [hunk 5]
```diff
@@ -156,6 +158,318 @@ public void UpdatePackageFromUI()
...
+            VisualStudio.ClearOutputWindow();
+            var uiwindow2 = nugetTestService.GetUIWindowfromProject(project);
+            uiwindow2.InstallPackageFromUI("contoso.a", "1.0.0");
+
+            // Assert
+            CommonUtility.AssertPackageInPackagesConfig(VisualStudio, project, "contoso.a", "1.0.0", XunitLogger);
+            CommonUtility.AssertPackageInPackagesConfig(VisualStudio, nuProject, "contoso.a", "1.0.0", XunitLogger);
+        }
+
+        [StaFact]
```
Comment (erdembayar) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs]: @nkolev92 
Here I was not able to add `InstallPackageFromUI_PackageNamespace_WithMultiFeed_Success()` test due to here unable to change source option to `All` with current Apex contract. Any change requires not only change in Apex contract also additional change (other to Apex test) in NuGet PMUI code is kind of out of scope of this PR.
Previously I thought NuGet PMUI searches from all sources for installing new package but it seems it's not true, if individual source is selected then it searches only from that source. It doesn't consider `Mypackages` for example in this case.
image
In order to make multifeed to work I need to select `All`, but currently not available due to previous mentioned limitation.

Comment (nkolev92) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs]: > . Any change requires not only change in Apex contract also additional change (other to Apex test) in NuGet PMUI code is kind of out of scope of this PR.

Not sure I agree here. 

We should add a test for this scenario. Apex test contracts are owned by us, it's equivalent to making a change to a type like: https://github.com/NuGet/NuGet.Client/blob/dev/test/TestUtilities/Test.Utility/Commands/TestRestoreRequest.cs

You can argue that this can be added in a new PR, but it's something that I'd say we should consider as part of the 6.0 MVP.

File: src/NuGet.Core/NuGet.Configuration/PackageNamespaces/PackageNamespacesConfiguration.cs [hunk 6]
```diff
@@ -38,7 +38,7 @@ public IReadOnlyList<string> GetConfiguredPackageSources(string term)
             return SearchTree.Value?.GetConfiguredPackageSources(term);
         }
 
-        internal PackageNamespacesConfiguration(Dictionary<string, IReadOnlyList<string>> namespaces)
+        public PackageNamespacesConfiguration(Dictionary<string, IReadOnlyList<string>> namespaces)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Configuration/PackageNamespaces/PackageNamespacesConfiguration.cs]: Are we making this public for test purposes?

I think you can use internalsvisibleto. 

I know in theory, it's not an ideal usage of the metadata, but that ship has long sailed, unless we aggressively make breaking changes.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGet.PackageManagement.Test.csproj [hunk 7]
```diff
@@ -11,6 +11,7 @@
   <ItemGroup>
     <ProjectReference Include="$(TestUtilitiesDirectory)Test.Utility\Test.Utility.csproj" />
     <ProjectReference Include="$(NuGetCoreSrcDirectory)NuGet.PackageManagement\NuGet.PackageManagement.csproj" />
+    <ProjectReference Include="..\NuGet.Configuration.Test\NuGet.Configuration.Test.csproj" />
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGet.PackageManagement.Test.csproj:14]: Why do we need this?

We tend to add common methods in the Test.Utility assembly.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs [hunk 8]
```diff
@@ -7034,6 +7035,230 @@ public async Task TestPacMan_PreviewInstallPackage_BuildIntegrated_NullVersion_T
...
+            <namespace id=""Contoso.*"" />             
+            <namespace id=""Test.*"" />
+        </packageSource>
+    </packageNamespaces>
+</configuration>");
+                var contosoPackageIdentity = new PackageIdentity("Contoso.A", new NuGetVersion("1.0.0"));
+
+                var ExternalA = new SimpleTestPackageContext()
+                {
+                    Id = contosoPackageIdentity.Id,  // Initial version had package id conflict with Contoso repository
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs]: Not sure I follow what this comment is implying.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs [hunk 9]
```diff
@@ -7034,6 +7035,230 @@ public async Task TestPacMan_PreviewInstallPackage_BuildIntegrated_NullVersion_T
             }
         }
 
+        [Fact]
+        public async Task TestPacManPreview_Install_PackageNamespace_Succeed()
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs]: Consider making this name more explicit. 

For example, this is testing a single package install with no dependencies.
There seems to be a namespace conflict but not sure that's super relevant. 

However, 0, 1 or more sources being selected is relevant.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs [hunk 10]
```diff
@@ -7034,6 +7035,230 @@ public async Task TestPacMan_PreviewInstallPackage_BuildIntegrated_NullVersion_T
             }
         }
 
+        [Fact]
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs:7038]: Given that the concept of primary and secondary sources exists, it's a good idea that we add some tests for that.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/ResolverGatherTests.cs [hunk 11]
```diff
@@ -1420,6 +1421,165 @@ public async Task ResolverGather_DependenciesSpreadAcrossRepos()
...
+
+            var providers3 = new List<Lazy<INuGetResourceProvider>>();
+            providers3.Add(new Lazy<INuGetResourceProvider>(() => new TestDependencyInfoProvider(packages3)));
+
+            var repos = new List<SourceRepository>();
+            repos.Add(new SourceRepository(new Configuration.PackageSource("http://1", "someRepository1"), providers1));
+            repos.Add(new SourceRepository(new Configuration.PackageSource("http://2", "someRepository2"), providers2));
+            repos.Add(new SourceRepository(new Configuration.PackageSource("http://3", configuredSources[0]), providers3));
+
+            var context = new GatherContext(namespacesConfiguration);
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.PackageManagement.Test/ResolverGatherTests.cs]: nit: If you want to improve readability consider switching to something like:

```cs
            var context = new GatherContext(namespacesConfiguration) 
            {
               PrimaryTargets = targets.ToList();
            } 

```

The quick actions are likely to suggest this.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs [hunk 12]
```diff
@@ -156,6 +158,318 @@ public void UpdatePackageFromUI()
             CommonUtility.AssertPackageInPackagesConfig(VisualStudio, project, "newtonsoft.json", "10.0.3", XunitLogger);
         }
 
+        [StaFact]
+        public async Task InstallPackageFromUI_PackageNamespace_WithSingleFeed_Match_Succeeds()
```
Comment (nkolev92) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs:162]: This name is a lot better than the ones in NuGetPackageManagerTests!

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs [hunk 13]
```diff
@@ -156,6 +158,318 @@ public void UpdatePackageFromUI()
...
+            var uiwindow2 = nugetTestService.GetUIWindowfromProject(project);
+            uiwindow2.InstallPackageFromUI("contoso.a", "1.0.0");
+
+            // Assert
+            CommonUtility.AssertPackageInPackagesConfig(VisualStudio, project, "contoso.a", "1.0.0", XunitLogger);
+            CommonUtility.AssertPackageInPackagesConfig(VisualStudio, nuProject, "contoso.a", "1.0.0", XunitLogger);
+        }
+
+        [StaFact]
+        public async Task InstallPackageToProjectsFromUI_PackageNamespace_WithSingleFeed_NoMatch_Succeeds()
```
Comment (nkolev92) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs]: The name rang a bell here, why would we succeed, if there's no match?
Shouldn't we fail in that case?

**Review by nkolev92 — COMMENTED** [review]
This is awesome progress.

I'm pretty impressed with how quickly you've been able to add these. 

Mostly some cleanup comments. 

I think you have identified, *the exact location* to make the change in, so that's great. 

Are you planning on adding tests for PMC & the ivs APIs in a future PR? 
Do the changes in this PR make it respected in PMC as well (I'm guessing yes is the answer here). 

It might be helpful to add some update tests as well. 
Note that adding those tests might mean adding tests to NuGetPackageManager that you are not necessarily testing.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs (hunk 5 above)
Comment (erdembayar) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs]: It can be done but changes would be kind of ugly, I have to change private methods/fields to public.
It only benefits Apex test but nothing else, so it would be to hard to justify it.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGet.PackageManagement.Test.csproj (hunk 7 above)
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGet.PackageManagement.Test.csproj:14]: I used `SettingsTestUtils.CreateConfigurationFile` method from `NuGet.Configuration.Test`. It already exist there, not new method.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs (hunk 5 above)
Comment (erdembayar) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs]: @nkolev92 
Can I create follow up issue for this Apex test case and do it separate PR?

Comment (nkolev92) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs]: > It can be done but changes would be kind of ugly, I have to change private methods/fields to public.
> It only benefits Apex test but nothing else, so it would be to hard to justify it.

Code in NuGet.Clients can be refactored as needed. 
Making the code testable would be a good investment. 

Do you expect this to be a very involved work?

> Can I create follow up issue for this Apex test case and do it separate PR?

Yep, I'm comfortable with that. In the meantime, please ensure the scenarios work manually (I know you've done that :) )

Comment (erdembayar) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs]: Ok. I created follow up issue https://github.com/NuGet/Home/issues/11026
Until this one is completed below 2 new unit tests technically cover it.

https://github.com/NuGet/NuGet.Client/blob/a8975bb6651dfbb4c50781895a2b48833304dcc8/test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs#L7117-L7121

https://github.com/NuGet/NuGet.Client/blob/a8975bb6651dfbb4c50781895a2b48833304dcc8/test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs#L7219-L7222

File: src/NuGet.Core/NuGet.Configuration/PackageNamespaces/PackageNamespacesConfiguration.cs (hunk 6 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.Configuration/PackageNamespaces/PackageNamespacesConfiguration.cs]: Ok. Addressed.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs (hunk 10 above)
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs:7038]: Added 1 test for it: `TestPacManPreview_InstallForPC_PackageNamespace_WithMultipleFeeds_SecondarySourcesNotConsidered`

https://github.com/NuGet/NuGet.Client/blob/dev-eryondon-NameSpacePackageConfig-InstallUpdate/test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs#L7218-L7321

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs (hunk 9 above)
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs]: Renamed all tests like `TestPacManPreview_InstallForPC_PackageNamespace_WithSingleFeed_Succeeds`, `TestPacManPreview_InstallForPC_PackageNamespace_WithMultipleFeedsWithIdenticalPackages_RestoresCorrectPackage` etc

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs (hunk 8 above)
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs]: Corrected it.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/ResolverGatherTests.cs (hunk 11 above)
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/ResolverGatherTests.cs]: Ok. Done.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs (hunk 12 above)
Comment (erdembayar) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs:162]: Ok. Improved names in `NuGetPackageManagerTests.cs`.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs (hunk 13 above)
Comment (erdembayar) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs]: Removed wrong assertion.

**Comment by erdembayar** [comment]
@nkolev92 
Thank you for review, I addressed latest PR comments. Please review again.

> Are you planning on adding tests for PMC & the ivs APIs in a future PR?
> Do the changes in this PR make it respected in PMC as well (I'm guessing yes is the answer here).

I'll create separate PR for PMC and IVs Api. Yes, this change is respected in PMC as well so next PR would be mostly End2End and Apex tests. I don't want too long PR.

> It might be helpful to add some update tests as well.
> Note that adding those tests might mean adding tests to NuGetPackageManager that you are not necessarily testing.

Since my [Apex tests are not complete](https://github.com/NuGet/Home/issues/11026) `NuGetPackageManager` tests technically cover scenarios for while. I guess more tests don't hurt, `NuGetPackageManager` tests are not heavy as integration tests and Apex tests.
Currently I have 1 update test and I'll create more with  https://github.com/NuGet/Home/issues/11026 https://github.com/NuGet/NuGet.Client/blob/a8975bb6651dfbb4c50781895a2b48833304dcc8/test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs#L318

File: src/NuGet.Core/NuGet.PackageManagement/Resolution/ResolverGather.cs [hunk 14]
```diff
@@ -31,6 +31,7 @@ public class ResolverGather
         private readonly HashSet<string> _idsSearched = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
         private int _maxDegreeOfParallelism;
         private readonly ConcurrentDictionary<string, TimeSpan> _timeTaken = new ConcurrentDictionary<string, TimeSpan>(StringComparer.OrdinalIgnoreCase);
+        private readonly bool _isPackageNamespaceEnabled;
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.PackageManagement/Resolution/ResolverGather.cs]: I imagine the name can follow the name of the other var: AreNamespacesEnabled

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs [hunk 15]
```diff
@@ -7034,6 +7035,402 @@ public async Task TestPacMan_PreviewInstallPackage_BuildIntegrated_NullVersion_T
...
+
+                // Act and Assert
+                var ex = await Assert.ThrowsAsync<InvalidOperationException>(async () =>
+                {
+                    await nuGetPackageManager.PreviewInstallPackageAsync(msBuildNuGetProject, contosoPackageIdentity.Id,
+                    new ResolutionContext(), new TestNuGetProjectContext(), primarySourceRepositoryProvider.GetRepositories(), secondarySourceRepositoryProvider.GetRepositories(), token);
+                });
+
+                // Even though Contoso.A 1.0.0 exist in ExternalRepository it wouldn't be picked for install.
+                // PrivateRepository is passed in secondary sources but for packages.config project not considered.
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs:7319]: There are scenarios where secondary sources are considered correct/ 

Can we add tests for that?

**Review by nkolev92 — COMMENTED** [review]
Looks great. 

A few more tests and 🚢

File: src/NuGet.Core/NuGet.PackageManagement/Resolution/ResolverGather.cs (hunk 14 above)
Comment (erdembayar) [reply src/NuGet.Core/NuGet.PackageManagement/Resolution/ResolverGather.cs]: Addressed.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs (hunk 15 above)
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs:7319]: @nkolev92 
I was not able to find other scenarios where secondary sources are considered. Please note in above test `secondary sources` is not incorrect, simply not searched(considered) from it. 
Could you be able to point what are other possible scenarios where secondary sources are considered?
I already checked update/downgrade action but it's actually just `uninstall(no source is considered)` + `install(above tested case)` action.

Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs:7319]: My recollection is that primary sources are considered for direct dependencies, while both primary and secondary are considered for transitive dependencies.

Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs:7319]: Ok. I added 2 more tests.

https://github.com/NuGet/NuGet.Client/blob/cd77b927c2e12d315113898ba8dc05801af48d60/test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs#L7324-L7325

https://github.com/NuGet/NuGet.Client/blob/cd77b927c2e12d315113898ba8dc05801af48d60/test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs#L7451-L7452

**Comment by erdembayar** [comment]
> Looks great.
> 
> A few more tests and 🚢

Thank you for your latest review. I added 2 more tests for `2ndary sources are considered for transitive dependencies install` and `2ndary sources are not considered update` scenario.
Please review again.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs [hunk 16]
```diff
@@ -7321,7 +7322,297 @@ public async Task TestPacManPreview_InstallForPC_PackageNamespace_WithMultipleFe
...
+
+                var ExternalDirectA = new SimpleTestPackageContext()
+                {
+                    Id = directPackageIdentity.Id, // Package Id not filtered by package namespace.
+                    Version = "1.0.0",
+                    Dependencies = new List<SimpleTestPackageContext>() { ExternalContosoA } // We set Contoso.A as dependent package.
+                };
+                ExternalDirectA.AddFile("lib/net461/directA.dll");
+
+                await SimpleTestPackageUtility.CreateFolderFeedV3Async(
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs:7379]: you are mkaing a similar call for both external direct and externalcontoso.

**Review by nkolev92 — APPROVED** [review]
LGTM!

Please ensure that there's tests for all the scenarios we have discussed. 
Hopefully the improved test names aid that :)

File: src/NuGet.Core/NuGet.Configuration/Properties/AssemblyInfo.cs [hunk 17]
```diff
@@ -11,4 +11,4 @@
...
 [assembly: InternalsVisibleTo("NuGet.Configuration.Test, PublicKey=0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9")]
 [assembly: InternalsVisibleTo("DynamicProxyGenAssembly2, PublicKey=0024000004800000940000000602000000240000525341310004000001000100c547cac37abd99c8db225ef2f6c8a3602f3b3606cc9891605d02baa56104f4cfc0734aa39b93bf7852f7d9266654753cc297e7d2edfe0bac1cdcf9f717241550e0a7b191195b7667bb4f64bcb8e2121380fd1d9d46ad2d92d2d15605093924cceaf74c4861eff62abf69b9291ed0a340e113be11e6a7d3113e92484cf7045cc7")]
-
+[assembly: InternalsVisibleTo("Test.Utility, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Configuration/Properties/AssemblyInfo.cs:14]: Why is this needed?  What internal API from `NuGet.Configuration` is `Test.Utility` using?

Comment (erdembayar) [reply src/NuGet.Core/NuGet.Configuration/Properties/AssemblyInfo.cs:14]: @zivkan 
It's used here.
https://github.com/NuGet/NuGet.Client/blob/cd77b927c2e12d315113898ba8dc05801af48d60/test/TestUtilities/Test.Utility/PackageNamespacesConfigurationUtility.cs#L35

Here is actual api:
https://github.com/NuGet/NuGet.Client/blob/9a72664c07aaef63e5d23a529e34f292f96e99d7/src/NuGet.Core/NuGet.Configuration/PackageNamespaces/PackageNamespacesConfiguration.cs#L41-L46

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs (hunk 16 above)
Comment (erdembayar) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/NuGetPackageManagerTests.cs:7379]: Ok. I didn't know we can create multiple nupkgs same time. I combined them into 1 call.

File: src/NuGet.Core/NuGet.Configuration/Properties/AssemblyInfo.cs (hunk 17 above)
Comment (zivkan) [reply src/NuGet.Core/NuGet.Configuration/Properties/AssemblyInfo.cs:14]: Wouldn't customers using our NuGet.Configuration package directly benefit from being able to use this API to mock configurations in their own tests, or even have hard-coded namespace configuration in their production applications, without having to read a file off disk?

Why isn't that API public?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Configuration/Properties/AssemblyInfo.cs:14]: I think we should err on the side of underexposing than overexposing APIs tbh. 

Given that modes are coming at a later point, it might be better to consider that later.
